### PR TITLE
Clarify the root project by printing "project : (the root project)" instead of just "project :" in resolution failure messaging

### PR DESCRIPTION
**Pull request by tresat**
Not a large or complex change, despite 31 files touched.

- Passing `Describable` instead of name String to Failures.
- When abstract Failures are asked for requested name, `ProjectPathClarifyingDescriber` is used to add description.
- Update test expectations.

New message example:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':forceResolution'.
> Could not resolve all dependencies for configuration ':resolveMe'.
   > Could not resolve project : (the root project).
     Required by:
         project : (the root project)
      > Multiple incompatible variants of org.example:xbdp7:1.0 were selected:
           - Variant blueElementsCapability1 has attributes {color=blue}
           - Variant greenElementsCapability2 has attributes {color=green}
```

Continues #27858

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test QFL

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Describes the given display path, clarifying when that path references the root project.
+     *
+     * @param displayPath the path to describe
+     * @return clarified description of the path
+     */
+    public static String describe(String displayPath) {
+        String pathAlone;
+        if (displayPath.startsWith(PROJECT_PREFIX)) {
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java:49]: While this works, I am a bit skeptical of playing with strings here. We have this data in its un-stringed form, and I think we might be able to handle this root project detection earlier

Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java:49]: I experimented a little with that, and wound up with ~2000 failing tests.  I don't want to fix all of those, or even analyze which of them need to be fixed.  I want to limit this change to resolution failure messages.

The implementations of `DefaultProjectComponentIdentifier` and `DefaultProjectComponentSelector` aren't very complex about how they build their display names.  While we could do an instanceof test and interpret the `ComponentIdentifier` and `ComponentSelector` types semantically, I don't think there is a need for that here.  Alternately, we could have these interfaces (or just their `DefaultProjectXYZ` implementations) each implement a new common interface and test for that.  These kind of seem like overkill for this problem.

I was kind of thinking of this type as just a Decorator around the existing project description.  This is already cleaner than some of the existing message formatting code, which is very much still string bashing, so I'm not too concerned.

Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java:49]: I was not proposing changing the way `ProjectComponentIdentifier/Selector` format their display names, but that we should avoid string manipulation like this if possible. 

> While we could do an instanceof test and interpret the ComponentIdentifier and ComponentSelector types semantically, I don't think there is a need for that here.

This is more so what I was proposing. Is this that much more complicated?

Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/ProjectPathClarifyingDescriber.java:49]: It's not that it's that much more complicated to implement, it's just that `instanceof` branching is typically considered an OOP code smell to be avoided.

I think there needs to be a stronger reason to go that route here than general dislike for conditioning on String contents.  This is meant as a decoration around the existing String, so I don't see a huge problem with processing it slightly first - it still tests the extracted path properly against the root path constant.

I could be talked into moving the whole testing part into `Path` as a new `isRootPath(String)` method that handles the prefix internally, but that's new public API.

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — CHANGES_REQUESTED** [review]
❌ This isn't the right approach.  We have a display name for Project instances, so we should be using that instead of working backwards from the path.  It's also inconsistent with other places we show a project's display name.  

Included builds also complicate this.  I think we end up only calling the root project of the root build "root project" and everywhere else is a project with a fully qualified path.

**Comment by cobexer** [comment]
For me `project : (the root project)` does not communicate the name/path of the root project but looks like a typo with a space before the `:` and it's also not consistent with how we show task names in the message below:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':forceResolution'.
> Could not resolve all dependencies for configuration ':resolveMe'.
   > Could not resolve project : (the root project).
     Required by:
         project : (the root project)
      > Multiple incompatible variants of org.example:xbdp7:1.0 were selected:
           - Variant blueElementsCapability1 has attributes {color=blue}
           - Variant greenElementsCapability2 has attributes {color=green}
```
Please also note that a space before a colon may be expected by French readers (see https://en.wikipedia.org/wiki/Colon_(punctuation)#Spacing_and_parentheses) which makes the `:` part of the message instead of something the reader should notice.


Should we quote the name/path of the project in the messages to make them more consistent?

Like in:
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':forceResolution'.
> Could not resolve all dependencies for configuration ':resolveMe'.
   > Could not resolve project ':' (the root project).
     Required by:
         project ':' (the root project)
      > Multiple incompatible variants of org.example:xbdp7:1.0 were selected:
           - Variant blueElementsCapability1 has attributes {color=blue}
           - Variant greenElementsCapability2 has attributes {color=green}
```

**Comment by tresat** [comment]
Replaced by https://github.com/gradle/gradle/pull/29667

**Pull request closed without merging**